### ISSUE TITLE
[JavaScript] Some problems require lists to be wrapped in a secondary list

### ISSUE THREAD

**Issue by BinaryFissionGames** [project-lovelace/lovelace-engine#85]
For instance, the Exponential Growth problem. This problem requires a list of values to be returned. However, if you return a raw list of values, as one would expect:
```
function exponential_growth(x0, k, dt, N) {
    let iterations = [x0];

    for(let i = 0; i < N; i++) {
        iterations.push(iterations[i] + iterations[i] * k * dt);
    }

    return iterations;
}
```

You get the following error:

```
Error: Traceback (most recent call last):
  File "/engine/engine/api.py", line 240, in on_post
    user_test_case = problem.ProblemTestCase(None, problem.INPUT_VARS, input_tuple, problem.OUTPUT_VARS, user_output)
  File "/engine/problems/test_case.py", line 39, in __init__
    output_tuple, len(output_tuple)))
ValueError: output_vars and output_tuple must be of the same length but output_vars=['x'] (len=1), output_tuple=(9.442968447496288, 9.442968447496288, 9.442968447496288, 9.442968447496288, 9.442968447496288, 9.442968447496288, 9.442968447496288, 9.442968447496288, 9.442968447496288, 9.442968447496288, 9.442968447496288, 9.442968447496288, 9.442968447496288, 9.442968447496288, 9.442968447496288) (len=15)
```

However, if you wrap the final list in another list:
```
function exponential_growth(x0, k, dt, N) {
    let iterations = [x0];

    for(let i = 0; i < N; i++) {
        iterations.push(iterations[i] + iterations[i] * k * dt);
    }

    return [iterations];
}
```

The answer is accepted. However, I believe the first snippet should be correct, and accepted by the engine.

**Comment by ali-ramadhan** [comment]
You're definitely right. Hmmm I think we're doing something wrong when reading the output user solution into Python.

For Javascript, user output is saved to JSON then passed to the engine which tries to extract the user output from JSON into Python. Will try to play around with the relevant logic and make sure your first solution works, probably somewhere here... https://github.com/project-lovelace/lovelace-engine/blob/4e3db1598974702f7a4aed585890f0d7f3bdd1d1/engine/run_js.py#L55-L60

**Comment by BinaryFissionGames** [comment]
So the vibe I get from this code is that there's an assumption that there's a fundamental difference between returning a list of values and returning multiple values, which I guess is kinda true for python (since python has lists & tuples), but JS (and JSON) only has lists, which kind of confuses the issue.

Looking into the code a bit, I see the same check for the Julia runner (which probably exhibits the same behaviour):  https://github.com/project-lovelace/lovelace-engine/blob/main/engine/run_jl.py#L66-L71

The two concepts (returning multiple variables, returning a list) need to be reconciled at some point in the code, because in many languages there's no concept of returning multiple variables and no way to differentiate it.

So, I think what might need to happen is the runners should return a list if there's a list in the returned JSON, and then somewhere (where the amount of output variables is known, maybe `test_case_solution_correct` in `test_case.py` in the lovelace-problems repository, or maybe (probably?) in `api.py` in this repo), convert the list to the proper tuple type if it's a list AND the problem requires multiple output variables.

**Comment by ali-ramadhan** [comment]
Yeah you bring up a good point about tuples vs. lists. It sounds like a potential solution could be for the engine to always wrap the JSON output in a list (so if the function returns 1 output, the JSON contains a list with 1 element).

The problem modules specify how many outputs there are for each problem like in https://github.com/project-lovelace/lovelace-problems/blob/5f04abe4deea50d6b1f2ba8fdace751632308353/problems/ada_lovelaces_note_g.py#L13 so from that the engine should know how many outputs to expect.

**Comment by ali-ramadhan** [comment]
Right so I guess fundamentally Javascript cannot be used to distinguish between a problem that returns two numbers like `[4, 5]` and a problem that returns a list like `[1, 2, 3]`.

Do you think we can use dictionaries to get around this ambiguity somehow? I guess instead of `[4, 5]` we would then have `{1: 4, 2: 5}` and instead of `[1, 2, 3]` it would be `{1: [1, 2, 3]}` so no longer ambiguous.

Will have to think about this a bit more to figure out a way where the user doesn't have to think about dictionaries...

**Comment by BinaryFissionGames** [comment]
Let me draft something up real quick - I agree the return itself is ambiguous, but in the context of a particular problem it's not ambiguous whether it's supposed to be a multiple return or not.

**Comment by BinaryFissionGames** [comment]
Fixed with merge of #96 -- Julia may have the same problem, will take a look in a different issue.

**Comment by BinaryFissionGames** [comment]
Tested with Julia - the glue code takes care of the issue for Julia, so the same issue isn't present
